### PR TITLE
Transition from defining a Railtie to becoming a Rails Engine

### PR DESCRIPTION
**Pull request by agrobbin**
Hi folks!!

## What did we change?

The gem is now integrating with Rails via `Rails::Engine` rather than `Rails::Railtie`, leveraging the autoloading of `app/models`.

## Why are we doing this?

When going through the process of upgrading an application that uses sidekiq_publisher to Ruby 3, we ran into a really weird situation that somehow was traced to the `SidekiqPublisher::Job` AR model.

Essentially, when booting the application, with the `SidekiqPublisher::Job` class included, the application fails to boot, hanging indefinitely in some kind of endless loop inside of native Ruby C code. Tracing things via `lldb` did not provide much more information into what the root cause is, so I went through the process of simply commenting out various gems (starting first with those that have C extensions), eventually removing sidekiq_publisher, when things all of a sudden started working. I then proceeded to do a similar thing with various parts of this gem, before temporarily removing `SidekiqPublisher::Job`, and having that turn out to be the culprit.

What is truly infuriating is that when attempting to do this on a fresh Rails 6.0.3.4 application (the version our application is currently on), this behavior does not occur, which I'd imagine means there is some interplay issues going on with some other gem/code.

I wish I had more specific details about why that class is an issue, but I have yet to reduce the issue down to a minimal reproducible case. What I *was* able to figure out is that removing the explicit `require 'activerecord'` and transitioning this gem to a Rails Engine resolves things, I believe due to the way Rails handles the order of loading models included in an engine.

I'd love to learn more about why this *actually* resolves the issues we've run into, but on the bright side, the resolution was pretty simple, and actually simplifies the `SidekiqPublisher::Job` class by not needing to `require 'activerecord'` explicitly, nor `require 'sidekiq_publisher/job'` itself.

## How was it tested?
- [x] Specs
- [x] Locally
- [x] Staging

### CODE REVIEW

File: lib/sidekiq_publisher.rb [hunk 1]
```diff
@@ -1,16 +1,15 @@
 # frozen_string_literal: true
 
-require "active_support"
+require "active_support/core_ext/numeric/time"
```
Comment (agrobbin) [inline lib/sidekiq_publisher.rb:3]: The `require "active_record"` in `sidekiq_publisher/job` was indirectly doing this for us. Now that we aren't doing that, we have to explicitly do this ourselves (it's used on line 16 below).

While I was figuring this out, I figured it was worth more specifically targeting what is needed from Active Support throughout the gem. If the preference is to keep the wholesale `require "active_support"`, just let me know!

Comment (tjwp) [reply lib/sidekiq_publisher.rb:3]: `require "active_support"` is not the wholesale require. It is documented (https://guides.rubyonrails.org/active_support_core_extensions.html) that you should call `require "active_support"` and then specific extensions. 

The require for everything would be `require "active_support/all"`.

Comment (agrobbin) [reply lib/sidekiq_publisher.rb:4]: @tjwp TIL! Just updated the first commit here to keep the `require "active_support"`, and then the specific extensions needed.

**Review by tjwp — APPROVED** [review]
This looks reasonable to me. I think this should be released as a new major version.

Could @ezcater/core-platform also take a look to see if they have any concerns?

**Review by kraila — APPROVED** [review]
Looks good to me

**Review by aprescott — APPROVED** [review]
I'm fine with this, although admittedly I don't have much insight into the functional differences between using an engine compared to a railtie for this. Seems like there aren't any than load ordering as @agrobbin mentioned? (May be worth us running this through some internal testing with a prerelease since we have a variety of apps with a variety of config.)

@agrobbin could you add a note to the CHANGELOG as an unreleased change just to capture it there?

**Comment by agrobbin** [comment]
@aprescott done!

**Comment by agrobbin** [comment]
We just ran into another situation that would benefit from this change. When upgrading to Rails 6.1, and setting the new framework defaults, we ran into a load order issue caused by when sidekiq_publisher loads Active Record.

When using the `rails app:update` command, `config/initializers/new_framework_defaults_6_1.rb` is created, which includes this:

```ruby
Rails.application.config.active_record.has_many_inversing = true
```

This is first stored in the app config options, and then set on `ActiveRecord::Base` [here](https://github.com/rails/rails/blob/v6.1.3.1/activerecord/lib/active_record/railtie.rb#L204-L212) as part of the app initialization process.

However, because sidekiq_publisher explicitly requires `active_record`, and then uses `ActiveRecord::Base`, it is loaded and initialized *before* `config/initializers/new_framework_defaults_6_1.rb` is called in the app initialization process, resulting in the following behavior:

```ruby
# without this change

Rails.configuration.active_record.has_many_inversing
# => true
ActiveRecord::Base.has_many_inversing
# => false

# with this change

Rails.configuration.active_record.has_many_inversing
# => true
ActiveRecord::Base.has_many_inversing
# => true
```

Essentially, while we did set the config, it had no effect because Active Record had already been loaded, and its configs already set.

It'd be awesome to get this merged to eliminate the out-of-order initialization!

**Comment by tjwp** [comment]
I'm going to go ahead and merge this and cut a pre-release. Even though this is not expected to be a breaking change. I plan to make this a 2.0 release as it is a significant change in the gem's implementation.

**Comment by agrobbin** [comment]
Thanks @tjwp!!

**Comment by halfwolf** [comment]
Thanks for following up Tim! Sorry for not following up @agrobbin it kept getting lost in a sea of github notifications/emails

**Comment by agrobbin** [comment]
@halfwolf no problem!!

**Pull request merged**